src/qcorrect/decorator.py

```diff
@@ -45,26 +51,47 @@ def wrapper(cls):
 
     return wrapper
 
-
 @custom_guppy_decorator
 def operation(defn):
-    guppy_dec = guppy.declare(defn)
-    func_ast, _ = parse_py_func(
-        DEF_STORE.raw_defs[guppy_dec.id].python_func, DEF_STORE.sources
-    )
-    ty = check_signature(func_ast, Globals(DEF_STORE.frames[guppy_dec.id]))
-
-    op_def = OpDef(
-        name=defn.__name__,
-        description="",
-        signature=OpDefSig(poly_func=ty.to_hugr_poly()),
-        lower_funcs=[],
-    )
-
-    hugr_ext.add_op_def(op_def)
-
-    def empty_dec() -> None: ...
-
-    return guppy.hugr_op(
-        quantum_op(defn.__name__, ext=hugr_ext), name=defn.__name__, signature=ty
-    )(empty_dec)
+    """Decorator to define code operations"""
+    # TODO: link operations to tket operations
+    # TODO: check signature of operation definition
+
+    defn.__setattr__("qct_op", True)
+
+    return defn
+
+class CodeDefinition:
+    # TODO: make this a frozen dataclass
+    frame: FrameType
+
+    def get_module(self):
+
+        self.guppy_module = ModuleType(self.__class__.__name__)
+        self.hugr_ext = Extension(self.__class__.__name__, SemanticVersion(0, 1, 0))
+
+        for name, defn in inspect.getmembers(self, predicate=inspect.ismethod):
+            if hasattr(defn, "qct_op"):
+                guppy_defn = defn().wrapped.python_func
+                func_ast, _ = parse_py_func(guppy_defn, DEF_STORE.sources)
+                ty = check_signature(func_ast, Globals(self.frame))
+
+                op_def = OpDef(
+                    name=defn.__name__,
+                    description=defn.__doc__ or "",
+                    signature=OpDefSig(poly_func=ty.to_hugr_poly()),
+                    lower_funcs=[], # TODO: Get lowering function
+                )
+
+                self.hugr_ext.add_op_def(op_def)
+
+                def empty_dec() -> None: ...
+
+                guppy_op = guppy.hugr_op(
+                    quantum_op(defn.__name__, ext=self.hugr_ext),
```

Comment:
`quantum_op` is an internal guppylang function so I'm not sure how stable it is, but I think we don't really need it -- can we just do
```suggestion
                    lambda ty, inst: ops.ExtOp(op_def, ty),
```
?

Reply:
yes I support using the hugr interface directly here rather than the guppy utility